--- client/third_party/breakpad/BUILD.gn	2024-06-11 15:04:34.411557138 +0000
+++ client/third_party/breakpad/BUILD.gn.new	2024-06-11 15:10:20.827239929 +0000
@@ -37,6 +37,7 @@
       "-Wno-non-c-typedef-for-linkage",  # typedef struct{} Options
       "-Wno-tautological-constant-out-of-range-compare",
       "-Wno-unused-parameter",  # argc of main in sym_upload.cc
+      "-Wno-deprecated-declarations",
     ]
   }
   if (is_win_gcc) {
@@ -56,7 +57,7 @@
   }
 
   if (is_clang) {
-    cflags = [ "-Wno-unused-parameter" ]
+    cflags = [ "-Wno-unused-parameter", "-Wno-deprecated-declarations" ]
   }
 }
 
