diff --git a/utils/bazel/configure.bzl b/utils/bazel/configure.bzl
--- a/utils/bazel/configure.bzl	(revision 82e851a407c52d65ce65e7aa58453127e67d42a0)
+++ b/utils/bazel/configure.bzl	(date 1705188246176)
@@ -32,7 +32,7 @@
 ]

 def _overlay_directories(repository_ctx):
-    src_path = repository_ctx.path(Label("@llvm-raw//:WORKSPACE")).dirname
+    src_path = repository_ctx.path(Label("@llvm//:WORKSPACE")).dirname
     bazel_path = src_path.get_child("utils").get_child("bazel")
     overlay_path = bazel_path.get_child("llvm-project-overlay")
     script_path = bazel_path.get_child("overlay_directories.py")
