diff --git a/utils/bazel/third_party_build/zstd.BUILD b/utils/bazel/third_party_build/zstd.BUILD
index 510c461390d8..cb6cab8c018c 100644
--- a/utils/bazel/third_party_build/zstd.BUILD
+++ b/utils/bazel/third_party_build/zstd.BUILD
@@ -16,7 +16,7 @@ bool_flag(

 config_setting(
     name = "llvm_zstd_enabled",
-    flag_values = {":llvm_enable_zstd": "true"},
+    flag_values = {":llvm_enable_zstd": "false"},
 )

 cc_library(
