diff --git a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
index acffa72124d8..66fe263dcef1 100644
--- a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
@@ -1292,10 +1292,17 @@ cc_library(
     ],
 )
 
+# Need to patch tooling_dependency_scanning
+# https://github.com/bazelbuild/rules_go/issues/2848
+# Not ideal, but can't compile binaries without this.
 cc_library(
     name = "tooling_dependency_scanning",
     srcs = glob(["lib/Tooling/DependencyScanning/**/*.cpp"]),
     hdrs = glob(["include/clang/Tooling/DependencyScanning/**/*.h"]),
+    linkstatic = select({
+        "@bazel_tools//tools/cpp:mingw": True,
+        "//conditions:default": False,
+    }),
     deps = [
         ":basic",
         ":codegen",
