diff --git a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
index 037719a51dd1..99a6942341fb 100644
--- a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
@@ -1484,6 +1484,20 @@ gentbl(
     ],
 )
 
+gentbl(
+    name = "driver_options_json_gen",
+    tbl_outs = [(
+        "",
+        "clang-options.json",
+    )],
+    tblgen = "//llvm:llvm-tblgen",
+    tblgen_args = "--dump-json",
+    td_file = "include/clang/Driver/Options.td",
+    td_srcs = [
+        "//llvm:include/llvm/Option/OptParser.td",
+    ],
+)
+
 cc_library(
     name = "driver",
     srcs = glob(
