diff --git a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
index 99a6942341fb..acffa72124d8 100644
--- a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
@@ -40,6 +40,13 @@ cc_binary(
     copts = [
         "$(STACK_FRAME_UNLIMITED)",
     ],
+    linkopts = select({
+        "@bazel_tools//src/conditions:windows": [
+            "-luuid",
+            "-lole32",
+        ],
+        "//conditions:default": ["-ldl"],
+    }),
     stamp = 0,
     deps = [
         ":support",
diff --git a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
index 9de9ad8ae2ff..848e86b57c4a 100644
--- a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
@@ -702,6 +702,13 @@ cc_binary(
         "include/llvm/TargetParser/SubtargetFeature.h",
     ]),
     copts = llvm_copts,
+    linkopts = select({
+        "@bazel_tools//src/conditions:windows": [
+            "-luuid",
+            "-lole32",
+        ],
+        "//conditions:default": [],
+    }),
     stamp = 0,
     deps = [
         ":CodeGenTypes",
diff --git a/utils/bazel/llvm-project-overlay/llvm/config.bzl b/utils/bazel/llvm-project-overlay/llvm/config.bzl
index b15ec9e1bb39..3e70b04cf193 100644
--- a/utils/bazel/llvm-project-overlay/llvm/config.bzl
+++ b/utils/bazel/llvm-project-overlay/llvm/config.bzl
@@ -75,6 +75,7 @@ win32_defines = [
     # LLVM features
     r'LTDL_SHLIB_EXT=\".dll\"',
     r'LLVM_PLUGIN_EXT=\".dll\"',
+    "HAVE_UNISTD_H=1",
 ]
 
 # TODO: We should switch to platforms-based config settings to make this easier
