diff --git a/go/private/rules/binary.bzl b/go/private/rules/binary.bzl
index e79cbda5..7b54682b 100644
--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -204,9 +204,12 @@ def _go_tool_binary_impl(ctx):
             output = bat,
             content = cmd,
         )
+        bargs = ctx.actions.args()
+        bargs.add("/c", bat.path.replace("/", "\\"))
         ctx.actions.run(
-            executable = bat,
-            inputs = sdk.libs + sdk.headers + sdk.tools + ctx.files.srcs + [sdk.go],
+            executable = "cmd",
+            arguments = [bargs],
+            inputs = sdk.libs + sdk.headers + sdk.tools + ctx.files.srcs + [sdk.go] + [bat],
             outputs = [cout],
             env = {"GOROOT": sdk.root_file.dirname},  # NOTE(#2005): avoid realpath in sandbox
             mnemonic = "GoToolchainBinaryCompile",
