diff --git a/go/private/rules/binary.bzl b/go/private/rules/binary.bzl
index 51530b04..88d2d0a9 100644
--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -442,9 +442,12 @@ def _go_tool_binary_impl(ctx):
             output = bat,
             content = cmd,
         )
+        bargs = ctx.actions.args()
+        bargs.add("/c", bat.path.replace("/", "\\"))
         ctx.actions.run(
-            executable = bat,
-            inputs = sdk.headers + sdk.tools + sdk.srcs + ctx.files.srcs + [sdk.go],
+            executable = "cmd",
+            arguments = [bargs],
+            inputs = sdk.headers + sdk.tools + sdk.srcs + ctx.files.srcs + [sdk.go] + [bat],
             outputs = [out, gopath, gocache],
             mnemonic = "GoToolchainBinaryBuild",
         )
