--- go/private/rules/binary.bzl	2024-06-04 15:30:43.511943305 +0000
+++ go/private/rules/binary.bzl.new	2024-06-04 15:30:30.391960478 +0000
@@ -461,9 +461,12 @@
             output = bat,
             content = cmd,
         )
+        bargs = ctx.actions.args()
+        bargs.add("/c", bat.path.replace("/", "\\"))
         ctx.actions.run(
-            executable = bat,
-            inputs = sdk.headers + sdk.tools + sdk.srcs + ctx.files.srcs + [sdk.go],
+            executable = "cmd",
+            arguments = [bargs],
+            inputs = sdk.headers + sdk.tools + sdk.srcs + ctx.files.srcs + [sdk.go] + [bat],
             outputs = [out, gotmp],
             mnemonic = "GoToolchainBinaryBuild",
         )
