--- go/private/rules/binary.bzl
+++ go/private/rules/binary.bzl.new
@@ -510,10 +510,13 @@ exit /b %GO_EXIT_CODE%
             output = bat,
             content = cmd,
         )
+        bargs = ctx.actions.args()
+        bargs.add("/c", bat.path.replace("/", "\\"))
         ctx.actions.run(
-            executable = bat,
+            executable = "cmd",
+            arguments = [bargs],
             inputs = depset(
-                ctx.files.srcs + [sdk.go],
+                ctx.files.srcs + [sdk.go] + [bat],
                 transitive = [sdk.headers, sdk.srcs, sdk.tools],
             ),
             outputs = [out, gotmp],
