--- go/private/rules/binary.bzl
+++ go/private/rules/binary.bzl.new
@@ -560,10 +560,13 @@
             output = bat,
             content = cmd,
         )
+        bargs = ctx.actions.args()
+        bargs.add("/c", bat.path.replace("/", "\\"))
         ctx.actions.run(
-            executable = bat,
+            executable = "cmd",
+            arguments = [bargs],
             tools = depset(
-                ctx.files.srcs + [sdk.go],
+                ctx.files.srcs + [sdk.go] + [bat],
                 transitive = [sdk.headers, sdk.srcs, sdk.tools],
             ),
             toolchain = None,
