--- build_defs/BUILD.bazel	2024-03-12 11:01:14.000000000 +0000
+++ build_defs/BUILD.bazel.new	2024-06-07 18:29:38.531498332 +0000
@@ -14,10 +14,24 @@
 )
 
 create_compiler_config_setting(
-    name = "config_msvc",
+    name = "config_msvc_cl",
     value = "msvc-cl",
 )
 
+# Caveat: clang-cl support in protobuf is only best-effort / untested for now.
+create_compiler_config_setting(
+    name = "config_clang_cl",
+    value = "clang-cl",
+)
+
+selects.config_setting_group(
+    name = "config_msvc",
+    match_any = [
+        ":config_clang_cl",
+        ":config_msvc_cl",
+    ],
+)
+
 config_setting(
     name = "aarch64",
     values = {"cpu": "linux-aarch_64"},
